Arduino/Firmata device support

pgrawehr opened the pull request:
Fixes #262 

This adds support for controlling an Arduino board directly from a PC. The arduino board will run a special firmware called "Firmata", that interprets the commands from the PC and set it to hardware. See Readme for details.

This also includeds an `ArduinoBoard` class, which is a prototype for #878. I'll later push more commits that will show how that would look like with the existing Windows/Linux drivers (but that's somehow also dependent on #1038, because things get complicated otherwise, see #1037 for one of the issues) 

@Ellerbach Can you check whether that interface would fit your FT4222 as well?

The first implementation for the firmata driver was from https://github.com/ms-iot/remote-wiring. MIT license.

joperezr commented:
NIT: I believe Arduino is kind of a very Generic Name to have this be the Arduino.csproj. Might get confusing people thinking that this is a full Arduino binding which is not really the case. Can we instead name it as Arduino.Firmata or something like that?

joperezr commented on src/devices/Arduino/Arduino.sln:
```diff
@@ -0,0 +1,93 @@
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Arduino.sample", "samples\Arduino.sample.csproj", "{2670F7BF-A7C8-49EB-9A99-1719A90D0C67}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Bmxx80", "..\Bmxx80\Bmxx80.csproj", "{EEEB0FB8-E1ED-4970-BDF6-DA3D5E2ED074}"
```
NIT: Do you need to include these projects? if they are just project references, you don't need to have them in the solution file.

pgrawehr replied:
Intellisense and Resharper are acting weird if the projects are not part of the solution. They fail to recognize the types from the referenced libraries if they're not part of the solution.

joperezr replied:
Yeah I’ve seen that happen too. The way to fix that is to open the settings in VS and disable nuget package restore of missing packages from the nuget options. Basically jn order for VS to understand how to use restore correctly it requires all referenced projects to be part of the solution, but if you disable restore from VS and just run restore from the command line before opening VS then VS will work as expected and won’t require the projects to be part of the solution.

joperezr commented on src/devices/Arduino/Arduino.csproj:
```diff
@@ -0,0 +1,29 @@
+    <TargetFramework>netcoreapp2.1</TargetFramework>
+    <EnableDefaultItems>false</EnableDefaultItems>
+    <RootNamespace>Iot.Device.Arduino</RootNamespace>
+    <Configurations>Windows-Debug;Linux-Debug;Linux-Release;Windows-Release;Debug;Release</Configurations>
```
For now lets keep consistency and don't define these Configurations or the Conditions on the Project references below. All of it is going to go away anyway very soon when we merge your other PR, and I plan on making a tool that will fix all projects so I want to have them as consistent as possible.

pgrawehr replied:
I agree. When we merge the other PR this is obsolete anyway. I'll remove this when the PR is ready to be merged (or clean it up if we merge the other one first).

joperezr commented on src/devices/Bmxx80/Bmxx80.csproj:
```diff
@@ -4,6 +4,7 @@
     <TargetFramework>netcoreapp2.1</TargetFramework>
     <EnableDefaultItems>false</EnableDefaultItems>
     <LangVersion>8.0</LangVersion>
+    <Configurations>Windows-Debug;Linux-Debug;Linux-Release;Windows-Release;Debug;Release</Configurations>
```
Same comment here and in the SoftwareSpi project around removing configurations and ProjectReference conditions.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,177 @@
+        private string _firmwareName;
+        private List<SupportedPinConfiguration> _supportedPinConfigurations;
+
+        public ArduinoBoard(Stream serialPortStream)
```
Should we also allow constructor overloads that take in either a SerialPort object or just an int for the port that is using?

pgrawehr replied:
I was thinking of doing that, yes, but I wanted to ask you whether it's ok to include a reference to System.IO.Ports. But now I see that this is part of Iot.Device.Bindings already. 
I'll add an overload that takes a port name (on linux, it may be /dev/ttyUSBx or /dev/serialx or /dev/ttyAMAx, so a number is not enough).

Ellerbach replied:
In the other bindings, it usually just take a string and then manage the serial port internally as yes, it can be any string.

pgrawehr replied:
I prefer to keep the Stream interface as well, because it's more generic. Maybe one wants to forward this over a network or replay data from a file.

Ellerbach replied:
Yes, sure, keep the stream and add the string to open as serial port then. And yes, some implementation exists over network.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,177 @@
+
+        public event Action<string, Exception> LogMessages;
+
+        public virtual void Initialize()
```
NIT: we usually call Initialize from constructor and make it private.

pgrawehr replied:
I know, but I'm not sure this is a good idea here, since this eventually needs to do a lot of stuff and can fail for various reasons. And I don't like throwing complex exceptions from constructors. 
This has several reasons:
- If an error happens, you have to handle them immediately and free resources in a finaly clause (i.e. if you open a serial port and then detect that the device is not there)
- It makes unit testing very hard, because you can't easily introduce Mocks.
- There's no way the client can attempt to continue after an error (i.e. the initialization fails, because a wrong firmware is on the device, then the client may go to error state but still have a device connection to maybe upload the correct fw).

joperezr replied:
That makes sense and I agree with you it’s best to keep it separate from the constructor. Can we then have a private field `bool _initialized` that we set to false until we call this method, and then make it so that every operation that requires initialization to first check if we need to call initialize and if so call it? What do you think about that?

pgrawehr replied:
Yes, that sounds fine.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,177 @@
+    /// Note that the program will run on the PC, so you cannot disconnect the
+    /// Arduino while this driver is connected.
+    /// </summary>
+    public class ArduinoBoard : IDisposable
```
Should this class be sealed?

pgrawehr replied:
I've (in parts) prepared to contain the virtual methods that will in the next step be moved to the `Board` class. After that, yes it probably should be sealed.

pgrawehr replied:
Actually, I think not. Maybe an ArduinoMegaSpecialBoard needs to override some methods.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,177 @@
+
+        public event Action<string, Exception> LogMessages;
+
+        public virtual void Initialize()
```
NIT: Why is this method virtual? do you expect to subclass ArduinoBoard and have subclasses redefine the method?

pgrawehr replied:
See above https://github.com/dotnet/iot/pull/1039#discussion_r407072894.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,177 @@
+            }
+        }
+
+        internal void Log(string message)
```
NIT: I know you are not using it, but perhaps you want this Log method to also have an exception parameter which is also passed in to the eventhandler

pgrawehr replied:
True, I'll fix that.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,177 @@
+
+        private void FirmataOnError(string message, Exception innerException)
+        {
+            LogMessages?.Invoke(message, innerException);
```
This should probably call Log()

Ellerbach replied:
For log, I've added a logger solution because I needed it for the NFC reader development. See https://github.com/dotnet/iot/blob/master/src/devices/Card/LogInfo.cs, I guess you can reuse the logger.

pgrawehr replied:
Sounds good, but I guess we should move it to the core then.

Ellerbach replied:
Agree. I let @joperezr comments on this

pgrawehr replied:
I think sorting out the logging stuff should be on a separate PR, so I'll leave as-is for now.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,177 @@
+            LogMessages?.Invoke(message, innerException);
+        }
+
+        public GpioController CreateGpioController(PinNumberingScheme pinNumberingScheme)
```
NIT: add a default pinNumberingScheme. We usually default to logical

pgrawehr replied:
I agree.

pgrawehr replied:
Added overload (I intend to provide a default PinNumberingScheme from the base class)

joperezr commented:
I’ll continue reviewing this weekend but since you wanted to know I definitely think we are on the right track here. I love how the API surface is minimal and very easy to use, and won’t require new knowledge as in the end you are really just exposing spi, i2c and gpio devices.

pgrawehr commented:
> 
> 
> NIT: I believe Arduino is kind of a very Generic Name to have this be the Arduino.csproj. Might get confusing people thinking that this is a full Arduino binding which is not really the case. Can we instead name it as Arduino.Firmata or something like that?

Maybe use "Firmata" as namespace and ArduinoXyz for the concrete class names?

joperezr commented on src/devices/Arduino/ArduinoGpioControllerDriver.cs:
```diff
@@ -0,0 +1,303 @@
+#pragma warning disable CS1591
+namespace Iot.Device.Arduino
+{
+    internal class ArduinoGpioControllerDriver : GpioDriver
```
NIT: Perhaps just name this ArduinoGpioDriver?

joperezr commented on src/devices/Arduino/ArduinoGpioControllerDriver.cs:
```diff
@@ -0,0 +1,303 @@
+#pragma warning disable CS1591
+namespace Iot.Device.Arduino
+{
+    internal class ArduinoGpioControllerDriver : GpioDriver
```
Also, do you envision only creating these from the ArduinoBoard class exclusively? I ask because most of our drivers are public today, so that people can instantiate them if they need to. I don't have good examples from the top of my head, but would it be possible for people to pluga device (like a LCD screen) and  use pins from both the Arduino board and the phyisical RPi? If so, you would probably want people to be able to create their own driver which would internally use a RPiDriver and ArduinoDriver to forward the calls from pins to the right driver. Does that make sense?

pgrawehr replied:
First part: Yes, this should only be created from the Board. All devices should only be created from the board, as then the board can handle common stuff. 
Second part: Hmmm... I see your example. But wouldn't it be better to merge the pins at the controller level then and create a MergingGpioController? The bindings take GpioController instances, so doing the mapping on that level seems more natural than abstracting a driver, which is really hardware-bound. This would mean to unseal GpioController (and check what needs to be made virtual).

pgrawehr replied:
Decision is to unseal the GpioController.

Ellerbach commented:
> Can you check whether that interface would fit your FT4222 as well?

@pgrawehr, it's a very complementary approach. Same philosophy: give something easy on a *"normal"* Windows or Mac or Linux an access to GPIO, SPI and I2C with the PWM flavor and more. 

You went for a similar approach with a specific core communication element (your ArduinoBoard and the specific Firma classes). And for specific GPIO, I2C and PWM, for the driver which is the approach I took as well for the FT4222. I'm just surprised that there is no SPI support in Firma.  Is there a way to get it somehow? That would be super nice as well.

Ellerbach commented on src/devices/Arduino/SupportedMode.cs:
```diff
@@ -0,0 +1,69 @@
+        /// <summary>
+        /// The pin supports digital input
+        /// </summary>
+        DIGITAL_INPUT = (0x00),
```
I know it's ok to have all uppercase names in case of registers. For this specific enum, I feel like we can benefit of having a normal naming.

krwq replied:
I agree with @Ellerbach here. This enum is public and it should follow the naming convention. The only exception for that we have in dotnet/runtime is enum with 100+ values where values from spec are hard to remember and people frequently copy & paste them

pgrawehr replied:
I agree. I'll change it.

Ellerbach commented on src/devices/Arduino/README.md:
```diff
@@ -0,0 +1,47 @@
+## Preparing your Arduino
+You need to upload a special sketch to the Arduino. This sketch implements the "Firmata-Protocol", a communication protocol that allows to remotely control all the inputs and outputs of the board. See https://github.com/firmata/protocol/blob/master/protocol.md for details. 
+
+The binding requires Firmata Version 2.6, which is implemented i.e. by the ConfigurableFirmata project. 
```
OK, so I understand now why the .ino is part of the project. Still what is specific to the .ino that is not part of the main Firmata project?

pgrawehr replied:
ConfigurableFirmata uses another top-level architecture than firmata. It uses an OO abstraction layer arount the firmata library that allows to selectively enable or disable features. See http://firmatabuilder.com/. 
Also, the code is much better structured and readable, which helps in understanding the command handling on the Arduino side if something doesn't work as expected.

pgrawehr commented on src/devices/Arduino/CommonFirmataFeatures.ino:
```diff
@@ -0,0 +1,102 @@
+/*
```
I included it for convenience. There's even a web service (FirmataBuilder) that generates the file, but you need to make sure to select the correct options there, or some things wouldn't work. So providing a file that works might cause fewer support issues.

pgrawehr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,177 @@
+            _firmata.Open(_serialPortStream);
+            _firmata.OnError += FirmataOnError;
+            var protocolVersion = _firmata.QueryFirmataVersion();
+            if (protocolVersion != _firmata.QuerySupportedFirmataVersion())
```
The main Firmata Repo is at 2.5, but ConfigurableFirmata (which I used) is 2.6. I think I'm not using any feature from 2.6 so far, though. But since both projects are not very active, I believe it should be Ok to expect the later.

pgrawehr replied:
Oh, but yes, I should use a >= instead.

pgrawehr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,177 @@
+        /// </summary>
+        /// <param name="settings">Spi Connection settings</param>
+        /// <returns></returns>
+        public SpiDevice CreateSpiDevice(SpiConnectionSettings settings)
```
There's an existing protocol proposal for SPI over Firmata, but ASFAIK no implementation yet. It would probably not be to hard to add that implementation to ConfigurableFirmata. 
I thought SoftwareSPI was better than nothing, but I did not test this in any way yet.

Ellerbach replied:
What is the license associated to this ino file? My guess is that regarding the type of license from the original Firmata repo, doesn't seems very compatible with this repo.

pgrawehr replied:
The license for ConfigurableFirmata is LGPL. So coding against it as a library is fine. [Firmata-Builder](https://github.com/firmata/firmata-builder) (the generator that created this main application file) has a MIT license. So re-distribution (even with changes) inside this repo is OK.

pgrawehr replied:
I just tested software SPI over firmata with an MCP3008. It does not really work 😞 . It manages to run the clock at ~1.25kHz, with a big skew, while it normally runs at 1Mhz. 
I can try whether I get it to run somehow, but for now, I think we should declare this an unsupported feature for Arduino.

Ellerbach replied:
I was unfortunately expecting it not to work :-( Have you try the native implementation of SPI in Firmata or just the software SPI?

pgrawehr replied:
The bad part: The problem is that Firmata has no SPI implementation at all. So far there's only a protocol proposal, but no implementation for it (except maybe a prototype for StandardFirmata). 
The good part: I've quickly written an implementation for that proposal yesterday and that _does_ work. The maintainer of the firmata repos is accepting PRs, even if he's not actively working on it right now. 
In either case: This will come down to query whether the Arduino has SPI support and use it in that case or throw if not. That's what I should do anyway, because even if an implementation exists, it could be disabled.

Ellerbach replied:
That's great news and that would make a huge difference. Native SPI is used a lot and very much needed for most of the bindings we have. Great 😊 Looking forward to try it 😊

pgrawehr commented on src/System.Device.Gpio/System/Device/I2c/I2cCommunicationException.cs:
```diff
@@ -0,0 +1,34 @@
+    /// <summary>
+    /// This exception is thrown on I2C communication errors
+    /// </summary>
+    public class I2cCommunicationException : IOException
```
Currently, we keep getting very unspecific exceptions from I2C problems, typically "Error 121", but also TimeoutExceptions and the like. So I thought it would be good to have something more specific (but then use it for all kinds of errors in I2C communication). But I have no strong oppinion on this one.

Ellerbach replied:
I originally thought like you and with the first implementation I realize we never really care about the details of the exception except for debug. Rest of the time, you just want the device to work and work and always work. So worst case scenario, in the main code, you trap all Exceptions by category with retry methods so your code never break. And tracking the IOException is more than largely enough.

pgrawehr commented on src/devices/Arduino/ArduinoAnalogControllerDriver.cs:
```diff
@@ -0,0 +1,133 @@
+            VoltageReference = 5.0;
+            // Number of the first analog pin. Serves for converting between logical A0-based pin numbers and digital pin numbers.
+            // The value of this is 14 for most arduinos.
+            var firstPin = _supportedPinConfigurations.FirstOrDefault(x => x.PinModes.Contains(SupportedMode.ANALOG_INPUT));
```
So you mean I should list all the analog pins and the count them?

Ellerbach replied:
Yes, or anything that won't block if you just 1 or as many as 20 analogic pins with some weird numbering.

pgrawehr commented on src/System.Device.Gpio/System/Device/Analog/AnalogController.cs:
```diff
@@ -0,0 +1,214 @@
+        /// </summary>
+        /// <param name="numberingScheme">The numbering scheme used to represent pins provided by the controller.</param>
+        /// <param name="driver">The driver that manages all of the pin operations for the controller.</param>
+        public AnalogController(PinNumberingScheme numberingScheme, AnalogControllerDriver driver)
```
This whole PR is also a preview/suggestion for #878 - so that's intentional. 
I'm not sure this should inherit from GpioController, since typically you can't change the pin modes and you can't write to the pins. So an implementation would typically have to throw lots of "NotSupportedExceptions" which is not so nice.

Ellerbach replied:
Writing to an analogic pin is placing the pin into a PWM mode. And you can always change the pin mode (up to the fact that a pin supports or not a specific analogic/digital/pul)

pgrawehr replied:
That's true. Firmata actually names PWM analog out (although technically that's something else, and the Arduino Due for instance has real analog out pins). But mixing PWM and analog in gets a bit confusing for boards that don't support either. 

I'm not sure what the best approach is here, but right now it looks like keeping analog in (and eventually out) separate from PWM and GPIO would keep the interface simpler and reduce the possibilities of a wrong API usage.

Ellerbach replied:
Fair. so debate still open for the *"Analog Controller"* or equivalent naming and link to anything existing.

pgrawehr commented on src/devices/Arduino/FirmataDevice.cs:
```diff
@@ -0,0 +1,826 @@
+        }
+
+        /// <summary>
+        /// Used where?
```
Whoever implemented the code I "borrowed" this from had confused sender and receiver part in some places. This is one of them - That's how the firmware send error messages, but the client never needs that.

krwq commented:
@pgrawehr what's the remaining work here aside from the merge conflicts?

pgrawehr commented:
> 
> 
> @pgrawehr what's the remaining work here aside from the merge conflicts?

I don't remember in detail, but one of the open points is how we go further with the board PR. This basically bases on that one (or actually, a slightly older proposal of it). So it needs to be clarified whether the concept matches our future strategy.

krwq commented:
@pgrawehr what specific pieces of board abstraction is must have for this PR to move forward?

pgrawehr commented:
@krwq : it's been a while now, so I'm not 100% sure and will need to check again. 
What's certainly needed is the unsealing of the GpioController, ideally also the abstract board class, but that could likely also be put in later (eventually with breaking changes though,  when different method names will be used)

Ellerbach commented:
> The proposed AnalogController can stay in Iot.Devices for now.

For the ADC we can as well get inspiration an try to align with the nanoFramework one: https://github.com/nanoframework/lib-Windows.Devices.Adc. There are good elements that we already discussed in the related issues #878 and #90.

Ellerbach commented:
> AnalogController is moved to System.Devices namespace

And that does remind me as well that there are quite some Analog sensors in GrovePi. Clearly soem real work on that would allow to reuse some of the sensors but as well having GrovePi being a proper GpioController and AnalogController at the same time in a more reusable way.

pgrawehr commented:
> 
> 
> > The proposed AnalogController can stay in Iot.Devices for now.
> 
> For the ADC we can as well get inspiration an try to align with the nanoFramework one: [nanoframework/lib-Windows.Devices.Adc](https://github.com/nanoframework/lib-Windows.Devices.Adc). There are good elements that we already discussed in the related issues #878 and #90.

Not sure I can fullfull all of those requirements just yet, but I'll try. The Nanoframwork API is conceptually a bit different than the approach here, but it is also pretty small.

Ellerbach commented:
> The Nanoframwork API is conceptually a bit different than the approach here, but it is also pretty small.

I'm not saying it's the best ;-) I'm saying that in the debate and the aim to find something great, we should look at it. And I'm perfectly find to see this going in without it. And later having a break change in this binding to integrate, the well though future AnalogicController (or AdcContreller whatever it will be called)

pgrawehr commented:
So, I hope I have addressed all known issues, so this can serve as a start. 

- Addressed the findings above (or so I think)
- I did not need to unseal the GpioController on this PR just yet, because it entirelly has its own drivers. But of course it will benefit from the reference counting capability. Or from the possibility to define a `MergingGpioController` that can abstract that pins actually come from different hardware components (ie you use an Arduino as Gpio expander to a Raspi and want to control some hardware than needs 30 digital I/O pins) - see one of the comments above. 
- I added the AnalogController suggestion to the Iot.Devices assembly, but with the namespace it will eventually get in the end, as to minimize possible breaking changes. I took the approach now discussed for the GPIO: A AnalogPin instance being created by `Open()` instead of providing integer pin numbers all over the place.
- The ArduinoBoard should become a subclass of the `Board`, which is not present in this PR. It's members are intended to be virtual later.
- The Readme needs more focus
- I've created a quite nice example which uses a small character display to show different computer values (CPU temp, CPU load, room temperature...), but this depends on PR #1238, so I'll add it back in later.

Ellerbach commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,323 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
```
And now you've added the 3 lines headers, it will change for a 2 lines one :-D

pgrawehr replied:
??? Is this the wrong header?

Ellerbach replied:
The new header is now just 2 lines see #1242 :-)
```csharp
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
```

Ellerbach commented on src/devices/Arduino/ArduinoGpioControllerDriver.cs:
```diff
@@ -0,0 +1,306 @@
+            _arduinoBoard.Firmata.WriteDigitalPin(pinNumber, value);
+        }
+
+        protected override WaitForEventResult WaitForEvent(int pinNumber, PinEventTypes eventTypes, CancellationToken cancellationToken)
```
Any idea of the performance of the events?

pgrawehr replied:
I didn't take explicit measurements (maybe I should), but probably not to bad, since they're really handled as interrupts and no polling from the client side is required here.

Ellerbach replied:
> since they're really handled as interrupts and no polling from the client side is required here

Then that's perfect, it means that you can't miss any. Otherwise I would have kind of argued about having them. Now, you can definitely keep them :-)

joperezr commented:
[Triage] This PR depends on the board abstraction PR https://github.com/dotnet/iot/pull/1128 so we will wait on that to get merged before we can take this one.

krwq commented:
@pgrawehr would be nice to get this PR merged, there seems to be a bit of build errors

pgrawehr commented:
@krwq: I'll rebase and fix any issues ASAP. However, as already stated above, if we merge this before the Board PR, there might later be some breaking changes (as we decide on the final names for the different methods of the main board class)

pgrawehr commented:
@joperezr Please have a look if you find time. There are a few things that should probably be improved later (i.e. extensibility), but as a start, I think this is good to go.

krwq commented on src/devices/Arduino/ArduinoAnalogController.cs:
```diff
@@ -0,0 +1,101 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
```
2-line copyright headers

krwq commented on src/devices/HardwareMonitor/OpenHardwareMonitor.cs:
```diff
@@ -177,6 +177,11 @@ public IList<Sensor> GetSensorList()
                     string? identifier = Convert.ToString(sensor["Identifier"]);
                     string? parent = Convert.ToString(sensor["Parent"]);
                     string? type = Convert.ToString(sensor["SensorType"]);
+                    if (string.IsNullOrEmpty(name) || string.IsNullOrEmpty(identifier))
```
this looks like unrelated change

pgrawehr replied:
True, this was a fix for a bug I found from the sample application. Shall I put that into a separate PR?

krwq replied:
preferable separate PR (the missing copyright headers are ok though as they don't change the product)

pgrawehr replied:
Was merged separtelly

krwq commented on src/devices/Common/CommonHelpers.csproj:
```diff
@@ -9,11 +9,12 @@
-	<Compile Include="Iot/Device/Graphics/*.cs" />
-	<Compile Include="System/Device/*.cs" />
-	<Compile Include="System/Device/Gpio/*.cs" />
+	  <Compile Include="Iot/Device/Common/*.cs" />
```
you seem to be mixing tabs and spaces here and overall indentation is messed up

krwq commented on src/devices/Arduino/tests/xunit.runner.json:
```diff
@@ -0,0 +1,4 @@
+{
+    "parallelizeAssembly": false,
+    "parallelizeTestCollections": false
+}
```
this shouldn't be part of the folder. You can achieve the same adding a custom attribute to one of the files in the test project

pgrawehr replied:
I was unable to get that to work as expected. It was always at least trying to run two test classes in parallel, which is also not working here.

krwq commented on src/devices/Arduino/ArduinoAnalogInputPin.cs:
```diff
@@ -0,0 +1,75 @@
+
+        public override void DisableAnalogValueChangedEvent()
+        {
+            _autoReportingReferenceCount -= 1;
+            if (_autoReportingReferenceCount == 0)
+            {
+                _board.Firmata.AnalogPinValueUpdated -= FirmataOnAnalogPinValueUpdated;
+            }
```
`_autoReportingReferenceCount` can go negative if user is not careful, you should guard against it and throw

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,421 @@
+        /// <param name="comPorts">List of com ports. See <see cref="GetSerialPortNames"/>.</param>
+        /// <param name="baudRates">List of baud rates to test. <see cref="CommonBaudRates"/>.</param>
+        /// <returns>A board, already open and initialized. Null if none was found.</returns>
+        public static ArduinoBoard? FindBoard(List<string> comPorts, List<int> baudRates)
+        {
+            foreach (var port in comPorts)
+            {
+                foreach (var baud in baudRates)
+                {
+                    ArduinoBoard? b = null;
+                    try
+                    {
+                        b = new ArduinoBoard(port, baud);
+                        b.Initialize();
+                        return b;
+                    }
+                    catch (Exception x) when (x is NotSupportedException || x is TimeoutException || x is IOException || x is UnauthorizedAccessException)
+                    {
+                        b?.Dispose();
+                    }
+                }
+            }
+
+            return null;
+        }
```
you should not open all ports and check if they can initialize, you might accidentally cause some action in the unrelated device.

pgrawehr replied:
True, there is a certain risk that this could happen, but how would you otherwise detect a board?

krwq replied:
Honestly, I'd not try to detect it over serial port (it kinda is the same scenario as if you tried to check all I2C addresses and checked if your device binding was able to connect to one of them)... The only way you can detect it reliably is if your device is USB showing as serial port and then getting some info from USB itself but not sure if that is worth the complexity...

krwq replied:
(with I2C you're likely to cause less likely to cause any harm though, with SerialPort there is almost nothing defining any universal format of data, most of the I2C at least uses registers and reads it more or less the same)

pgrawehr replied:
Well, this is an optional feature (you do not have to use this auto-detect feature) and it works quite nicely for testing and you can specify the list of ports to scan (or ommit one that causes trouble). It is also very helpful since when connecting and disconnecting Arduinos (or generally USB-to-Serial devices) they often tend to change the port numbers. I don't want to change the code each time I reconnect a board.

pgrawehr replied:
Added corresponding warning in documentation.

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,421 @@
+        {
+            get
+            {
+                return _supportedPinConfigurations;
```
you should not return original mutable objects in the properties. Either use a copy or use some non-mutable type.

pgrawehr replied:
Good catch!

krwq commented on src/devices/CharacterLcd/LcdCharacterEncodingFactory.cs:
```diff
@@ -315,15 +315,22 @@ static LcdCharacterEncodingFactory()
                 { 'Χ', (byte)'X' },
             };
 
+            void AddToAllMaps(char c, char c1)
```
this looks like unrelated change

krwq commented on src/devices/Arduino/tests/FirmataTestFixture.cs:
```diff
@@ -0,0 +1,73 @@
+﻿using System;
```
missing copyright header

krwq commented on src/devices/Arduino/samples/DebugLogStream.cs:
```diff
@@ -0,0 +1,86 @@
+using System.IO;
+using System.Text;
+
+#pragma warning disable CS1591
```
please add a comment next to this what this warning is

krwq commented on src/devices/Bmxx80/ReadResult/Bme280ReadResult.cs:
```diff
@@ -1,6 +1,5 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
-// See the LICENSE file in the project root for more information.
```
I'm ok with this being in this PR only because this technically speaking doesn't change the product

krwq commented on src/devices/Arduino/samples/CharacterDisplay.cs:
```diff
@@ -0,0 +1,56 @@
+﻿using System;
```
copyright header

krwq commented on src/devices/Arduino/category.txt:
```diff
@@ -0,0 +1,6 @@
+i2c
```
usb?

krwq commented on src/devices/Arduino/category.txt:
```diff
@@ -0,0 +1,6 @@
+i2c
+usb
+gpio
+adc
+protocol
+spi
```
ok as is in this PR, I might change this though when regenerating the listing - the reason for this file is mostly so the listing looks coherent, there is so many categories here I might think of different category for this

pgrawehr replied:
I agree that one common name for such extension boards would be helpful, but I can't currently think of a good name. "GpioExpander" is not fitting, because it's more than just GPIO.

pgrawehr replied:
I guess this needs cleaning up once we have a matching category for such boards. The FT4222 also has the category USB, although it is conceptionally very similar to this binding.

krwq commented on src/devices/Arduino/README.md:
```diff
@@ -0,0 +1,90 @@
+
+## Known limitations
+
+All communication is routed trough the USB cable immitating a serial port with a limited bandwith. Therefore, some not insignificant delays are to be expected when sending commands or retrieving data. Communicating with sensors which have time-critical behavior will most likely not work reliably for this reason and the standard bindings provided for these won't work. This includes sensors like th ...
```
If this is all USB interaction then `usb` category should not be here. It's more serial port related than USB. USB is incidental here.

pgrawehr replied:
See above

krwq commented on src/devices/Arduino/FirmataCommand.cs:
```diff
@@ -0,0 +1,62 @@
+    /// <summary>
+    /// Primary firmata commands
+    /// </summary>
+    public enum FirmataCommand : byte
```
Does this have to be public? If so please user regular casing

krwq replied:
(honestly this looks like it shouldn't be public)

pgrawehr replied:
True, this was public by mistake.

krwq commented on src/devices/Arduino/ConfigurableFirmata/ConfigurableFirmata.ino:
```diff
@@ -0,0 +1,223 @@
+/*
+ * CommonFirmataFeatures.ino generated by FirmataBuilder
```
I assume this isn't part of the project. Make sure this is highly visible in the docs. Consider moving this into sample.

@richlander does externally generated file need to have copyright header?

pgrawehr replied:
This file (or some adaption of it) is used to load the arduino. If the Arduino IDE is installed, double clicking on this file opens up the IDE and directly allows flashing the firmware. It needs to be in a folder with the same name. 

I'll check again that this process is correctly documented.

krwq replied:
I understand, what I was asking if the more appropriate place to put this would be samples code, it's more likely to be noticed in there IMO

pgrawehr replied:
It's linked to in the Readme, so I hope it can be found.

krwq commented on src/devices/Arduino/ArduinoPwmChannel.cs:
```diff
@@ -0,0 +1,135 @@
+
+        private void Update()
+        {
+            if (_enabled)
+            {
+                _board.Firmata.SetPwmChannel(_pin, _dutyCycle);
+            }
+            else
+            {
+                _board.Firmata.SetPwmChannel(_pin, 0);
+            }
```
```suggestion
            _board.Firmata.SetPwmChannel(_pin, _enabled ? _dutyCycle : 0);
```

krwq commented on src/devices/Arduino/ArduinoI2cDevice.cs:
```diff
@@ -0,0 +1,102 @@
+
+        public ArduinoI2cDevice(ArduinoBoard board, I2cConnectionSettings connectionSettings)
+        {
+            _board = board ?? throw new ArgumentNullException(nameof(board));
```
no need to do all validation considering this is internal class. Just make sure you never pass invalid values. Note this is optional feedback - more validation doesn't hurt.

krwq commented on src/devices/Arduino/ArduinoAnalogInputPin.cs:
```diff
@@ -0,0 +1,79 @@
+            _configuration = configuration;
+        }
+
+        public override void EnableAnalogValueChangedEvent(GpioController? masterController, int masterPin)
```
I'm a bit confused on what this does and when it's used. Why not have event as part of this class?

pgrawehr replied:
This event is the link to the underlying transport connection and therefore binding-specific. The public event that you're looking for is in the base class. Forwarding takes place further down in `FirmataOnAnalogPinValueUpdated`. The base class has no direct knowledge about the source of the event.

krwq commented on src/devices/Arduino/ArduinoAnalogInputPin.cs:
```diff
@@ -0,0 +1,79 @@
+            _autoReportingReferenceCount += 1;
+        }
+
+        public override void DisableAnalogValueChangedEvent()
```
similar question as above

krwq commented on src/devices/Arduino/ArduinoAnalogInputPin.cs:
```diff
@@ -0,0 +1,79 @@
+            return _board.Firmata.GetAnalogRawValue(PinNumber);
+        }
+
+        public override bool SupportsAnalogInput()
```
this is really confusing to me. You're inside AnalogInputPin class, how could this ever return false?

krwq replied:
(also my gut tells me the more appropriate name would be: IsAnalogInput)

pgrawehr replied:
You are right, this is a leftover. Whether a certain pin supports analog input is tested long before we get here.

krwq commented on src/devices/Arduino/ArduinoAnalogInputPin.cs:
```diff
@@ -0,0 +1,79 @@
+            }
+        }
+
+        public override void QueryResolution(out int numberOfBits, out double minVoltage, out double maxVoltage)
```
could these be 3 properties instead?

pgrawehr replied:
Yes, not sure why I used this unconventional approach.

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,421 @@
+        /// <summary>
+        /// Default pin numbering scheme of this board. Always <see cref="PinNumberingScheme.Logical"/>, because the Arduino has only one pin numbering scheme.
+        /// </summary>
+        public PinNumberingScheme DefaultPinNumberingScheme
```
isn't Logical always the default scheme?

pgrawehr replied:
No, you can instantiate a board (well, not an arduino board, but in general) to use another scheme by default. This would make all pin numbers of all methods to use that scheme (like it is already done in the GpioController)

pgrawehr replied:
For this board, yes, but not in general. This property will later be overriding an abstract base property.

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,421 @@
+        /// <summary>
+        /// Attach to this event to retrieve log messages
+        /// </summary>
+        public event Action<string, Exception?>? LogMessages;
```
I assume that when Exception is null it means Info message? If so please document it as such or explain why Exception can be null

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,421 @@
+                    }
+                    catch (Exception x) when (x is NotSupportedException || x is TimeoutException || x is IOException || x is UnauthorizedAccessException)
+                    {
+                        b?.Dispose();
```
should you dispose on any exception?

pgrawehr replied:
There shouldn't be any other exceptions than these. If there are, it's most likely a programming error (i.e. a NullReferenceException) when I really don't care.

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,421 @@
+        /// <param name="comPorts">List of com ports. See <see cref="GetSerialPortNames"/>.</param>
+        /// <param name="baudRates">List of baud rates to test. <see cref="CommonBaudRates"/>.</param>
+        /// <returns>A board, already open and initialized. Null if none was found.</returns>
+        public static ArduinoBoard? FindBoard(List<string> comPorts, List<int> baudRates)
```
consider

```suggestion
        public static ArduinoBoard? FindBoard(IEnumerable<string> comPorts, IEnumerable<int> baudRates)
```

in the example just pass SerialPort.GetPortNames().

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,421 @@
+        /// </summary>
+        /// <returns>A list of available serial ports</returns>
+        /// <exception cref="Win32Exception">There was an error retrieving the list</exception>
+        public static List<string> GetSerialPortNames()
+        {
+            return SerialPort.GetPortNames().ToList();
+        }
```
see my suggestion to use IEnumerable, with that this method is not needed as you can pass SerialPort.GetPortNames() directly.

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,421 @@
+        {
+            get
+            {
+                return _firmwareVersion ?? new Version();
```
you can move this logic to the constructor and get rid of private field

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,421 @@
+            {
+                if (_protocolVersion == null)
+                {
+                    return new Version();
```
should we throw when we don't know protocol version?

pgrawehr replied:
That happens in the Initialize function.

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,421 @@
+        /// <returns>An instance of GpioController, using an Arduino-Enabled driver</returns>
+        public GpioController CreateGpioController()
+        {
+            return CreateGpioController(DefaultPinNumberingScheme);
```
I wouldn't create property DefaultPinNumberingScheme just so that you can have parameterless CreateGpioController

pgrawehr replied:
This is in preparation of the behavior of the (future) board base class. There you'll define the DefaultPinNumberingScheme in the ctor, but not change it later.

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,421 @@
+        /// <returns>An <see cref="I2cDevice"/> instance</returns>
+        /// <exception cref="NotSupportedException">The firmware reports that no pins are available for I2C. Check whether the I2C module is enabled in Firmata.
+        /// Or: An invalid Bus Id or device Id was specified</exception>
+        public virtual I2cDevice CreateI2cDevice(I2cConnectionSettings connectionSettings)
```
should ArduinoBoard have I2cBus as property or create bus instead of devices? I know currently only bus 0 is supported but:
- assume I'd want to combine the board with SenseHat or similar device - I can't do that without implementing bus myself
- what if new Arduino shows up with 2 I2cBuses

pgrawehr replied:
Some more expensive boards actually do have 2 busses (i.e. the Arduino Due), but for now the protocol doesn't support it. So this would also require a (potentially breaking) protocol extension. I don't really mind doing that (basically requiring our binding to require a specifically modified firmware), but it's out of scope here. 

About the Bus/Device destinction I think we should wait for the board PR here, as there will be some API adjustments like that anyway.

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,421 @@
+            return new ArduinoAnalogController(this, SupportedPinConfigurations, PinNumberingScheme.Logical);
+        }
+
+        /// <summary>
+        /// Special function to read DHT sensor, if supported
+        /// </summary>
+        /// <param name="pinNumber">Pin Number</param>
+        /// <param name="dhtType">Type of DHT Sensor: 11 = DHT11, 22 = DHT22, etc.</param>
+        /// <param name="temperature">Temperature</param>
+        /// <param name="humidity">Relative humidity</param>
+        /// <returns>True on success, false otherwise</returns>
+        public bool TryReadDht(int pinNumber, int dhtType, out Temperature temperature, out RelativeHumidity humidity)
+        {
+            if (!_supportedPinConfigurations[pinNumber].PinModes.Contains(SupportedMode.DHT))
+            {
+                temperature = default;
+                humidity = default;
+                return false;
+            }
+
+            return Firmata.TryReadDht(pinNumber, dhtType, out temperature, out humidity);
+        }
```
is DHT actually part of the spec? why is DHT special enough to be here?

pgrawehr replied:
There's a proposed spec for this (I'm still waiting on getting the rights to bring the specs to the next generation). It's not in the main firmware, but the advanced firmware (see Readme) supports it. 

The reason this needs a special command (and firmware module) is that this sensor's timing is so critical, that it doesn't work when using "our" binding and a default GPIO pin. See "Limitations" in Readme.

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,421 @@
+        protected virtual void Dispose(bool disposing)
+        {
+            // Do this first, to force any blocking read operations to end
+            if (_dataStream != null)
+            {
+                _dataStream.Close();
+                _dataStream.Dispose();
+                _dataStream = null!;
+            }
```
1. Why do you both Close and Dispose
2. Should you only do this when SerialPort is null? (nulling stream can happen regardless)

pgrawehr replied:
1. is likely a leftover from investigating problems with closing the port. 
2. It's more consistent this way and I don't want to touch this unless required, because of the risk running again into deadlocks when shutting everything down.

krwq replied:
What do you mean it's more consistent?

pgrawehr replied:
Ah sorry, I was unclear. I need to close the stream in either case, because that will terminate the blocking read in the receive thread. So yes, it is currently not possible to keep the stream open, even if it wasn't created from the Serial Port. But getting that to work would probably be really not worth the effort.

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,421 @@
+            if (_firmata != null)
+            {
+                _firmata.OnError -= FirmataOnError;
+                _firmata.Close();
+                _firmata.Dispose();
```
why isn't Diposing also closing? what's the difference?

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,421 @@
+            _spiEnabled++;
+            if (_spiEnabled == 1)
+            {
+                _firmata?.EnableSpi();
```
can you always enable it? Also might make sense to check for negative ref count

pgrawehr replied:
No, since that blocks the pins for other uses (if you don't need SPI, you can use the pins for GPIO).

krwq commented on src/devices/Arduino/ArduinoGpioControllerDriver.cs:
```diff
@@ -0,0 +1,305 @@
+
+        protected override void ClosePin(int pinNumber)
+        {
+            _pinModes.TryRemove(pinNumber, out _);
```
you're removing something on ClosePin but do nothing on OpenPin, that sounds like some inconsistency but didn't dig

pgrawehr replied:
The Add() is on `SetPinMode`, which is always called after `Open()`.

krwq commented on src/devices/Arduino/ArduinoGpioControllerDriver.cs:
```diff
@@ -0,0 +1,305 @@
+            PinMode ret;
+            switch (mode)
+            {
+                case SupportedMode.DIGITAL_OUTPUT:
```
can a pin be digital or analog? if so should you be checking for flags rather than value?

pgrawehr replied:
This is the current mode of a pin (as opposed to the pin capabilities). This can have one value at a time only.

krwq commented on src/devices/Arduino/ArduinoI2cDevice.cs:
```diff
@@ -0,0 +1,102 @@
+
+        public override byte ReadByte()
+        {
+            Span<byte> bytes = stackalloc byte[1];
+            bytes[0] = 0;
+            _board.Firmata.WriteReadI2cData(ConnectionSettings.DeviceAddress, null, bytes);
+            return bytes[0];
```
isn't this a default implementation (if you do not implement this wouldn't it do the same thing?)

pgrawehr replied:
No, the base implementation is abstract.

krwq commented on src/devices/Arduino/ArduinoI2cDevice.cs:
```diff
@@ -0,0 +1,102 @@
+            _board.Firmata.WriteReadI2cData(ConnectionSettings.DeviceAddress, null, buffer);
+        }
+
+        public override void WriteByte(byte value)
+        {
+            Span<byte> bytes = stackalloc byte[1];
+            bytes[0] = value;
+            _board.Firmata.WriteReadI2cData(ConnectionSettings.DeviceAddress, bytes, null);
+        }
```
same question

krwq commented on src/devices/Arduino/ArduinoPwmChannel.cs:
```diff
@@ -0,0 +1,128 @@
+
+            Channel = channel;
+            _pin = channel;
+            var caps = board.SupportedPinConfigurations.FirstOrDefault(x => x.Pin == _pin);
```
should you check for both mode and pin number here? is it possible that you find something with matching pin but different mode so the below check fails but there will be another pin which matches this?

pgrawehr replied:
No. The `SupportedPinConfigurations` structure contains the possible modes for each pin. Here, we get the set of modes that particular `_pin` supports. If it does not support PWM, we throw.

krwq commented on src/devices/Arduino/ArduinoPwmChannel.cs:
```diff
@@ -0,0 +1,128 @@
+            _frequency = frequency;
+            _dutyCycle = dutyCyclePercentage;
+            _board = board;
+            _enabled = false;
```
I'd initialize this in the field itself

krwq commented on src/devices/Arduino/FirmataDevice.cs:
```diff
@@ -0,0 +1,1234 @@
+    /// <summary>
+    /// This delegate is used to fire digital value changed events.
+    /// </summary>
+    public delegate void DigitalPinValueChanged(int pin, PinValue newValue);
```
Could this use https://github.com/dotnet/iot/blob/b5bb684bbf00ebd266018cfffde3dcbde26f0538/src/System.Device.Gpio/System/Device/Gpio/PinChangeEventHandler.cs instead?

pgrawehr replied:
Done. This is internal only, anyway.

krwq commented on src/devices/Arduino/FirmataDevice.cs:
```diff
@@ -0,0 +1,1234 @@
+            }
+        }
+
+        public void Close()
```
Why not move this logic to Dispose?

krwq commented on src/devices/Arduino/FirmataDevice.cs:
```diff
@@ -0,0 +1,1234 @@
+            }
+        }
+
+        public void Open(Stream stream)
```
should this be part of the constructor or static Create method?

pgrawehr replied:
I prefer to have these separated, for better testability (unit testing is simpler if ctors don't automatically start threads etc)

krwq commented on src/devices/Arduino/FirmataDevice.cs:
```diff
@@ -0,0 +1,1234 @@
+        private string PrintfFromByteStream(string fmt, in Span<byte> bytesReceived, int startOfArguments)
+        {
+            string output = fmt;
+            while (output.Contains("%"))
```
you might benefit from using StreamReader here

pgrawehr replied:
Doesn't work, at least not out of the box (I'm kind of "in-place" updating the string). Should probably be using StringBuilder, but several methods are missing there (i.e. there's no StringBuilder.Contains) and converting back and forth doesn't really help then as well.

krwq commented on src/devices/Arduino/samples/Arduino.Monitor.cs:
```diff
@@ -0,0 +1,300 @@
+        /// <summary>
+        /// Main entry point
+        /// </summary>
+        /// <param name="args">Unused</param>
```
unused? you use it below

krwq commented on src/devices/Arduino/samples/Arduino.Monitor.cs:
```diff
@@ -0,0 +1,300 @@
+            string portName = "COM4";
+            if (args.Length > 1)
+            {
+                portName = args[1];
```
why start with first argument not 0?

krwq commented on src/devices/Common/System/Device/Analog/ValueChangedEventArgs.cs:
```diff
@@ -0,0 +1,57 @@
+        /// <param name="newValue">The analog sensor reading, converted to voltage.</param>
+        /// <param name="pinNumber">The pin number that triggered the event.</param>
+        /// <param name="triggerReason">The reason for the event</param>
+        public ValueChangedEventArgs(uint rawValue, ElectricPotential newValue, int pinNumber, TriggerReason triggerReason)
```
I'd suggest changing the name of `newValue` to `value` or `electricPotential` and make sure property matches this name

krwq commented on src/devices/Common/System/Device/Analog/ValueChangedEventArgs.cs:
```diff
@@ -0,0 +1,57 @@
+        /// <summary>
+        /// Raw value, unscaled.
+        /// </summary>
+        public uint RawValue
```
I'm a bit sceptical of this being 32-bit value. For most ADCs this will work but I'm not sure this will really scale that well. Also note that RawValue is meaningless unless you also have MaxValue

pgrawehr replied:
The MaxValue is a property of the pin, so I'm not sure we should provide it basically as a duplicate. Some people might want the raw value (for whatever reason), so I think removing it wouldn't be helpful.

krwq commented on src/devices/Common/System/Device/Analog/TriggerReason.cs:
```diff
@@ -0,0 +1,35 @@
+        /// <summary>
+        /// A value is provided when certain thresholds are exceeded
+        /// </summary>
+        LimitExceeded
```
consider calling this `overflow`

pgrawehr replied:
The "overflow" can be either direction, and overflow would indicate the value is to high (when it can also become to small)

krwq commented on src/devices/Common/System/Device/Analog/TriggerReason.cs:
```diff
@@ -0,0 +1,35 @@
+        /// <summary>
+        /// The reason for the new message is unknown
+        /// </summary>
+        Unknown,
```
do we use it anywhere in the current code? I'd rather have us add new values when we need to rather than having generic reason on the list.

Also to me this seems more like Error enum. It's either OK reading or something went wrong.

pgrawehr replied:
This is considering future implementations of this interface/enum. Some ADCs have configurable interrupt capabilities and will fire an update only when the value changed in a specific way (i.e. it exceeds a certain threshold). Therefore I added this.

krwq commented on src/devices/Common/System/Device/Analog/AnalogInputPin.cs:
```diff
@@ -0,0 +1,179 @@
+            PinNumber = pinNumber;
+        }
+
+                /// <summary>
```
indent is off

krwq commented on src/devices/Common/System/Device/Analog/AnalogInputPin.cs:
```diff
@@ -0,0 +1,179 @@
+        /// <summary>
+        /// Close the pin
+        /// </summary>
+        public virtual void Close()
```
this shouldn't be public. The only reason we have both Close and Dispose on the Stream or XML is because it was a misdesign - Dispose didn't exist at that time and was added later and Close methods are there for legacy reason. There is no reason to have that in the new code. They semantically mean the same thing.

krwq commented on src/devices/Common/System/Device/Analog/AnalogController.cs:
```diff
@@ -0,0 +1,153 @@
+        /// Closes the given pin
+        /// </summary>
+        /// <param name="pin">The pin to close</param>
+        public virtual void Close(AnalogInputPin pin)
```
ClosePin/CloseAnalogPin - make sure to match the name with OpenPin/OpenAnalogPin

pgrawehr replied:
Yea, calling it `ClosePin`.

krwq commented on src/devices/Common/System/Device/Analog/AnalogController.cs:
```diff
@@ -0,0 +1,153 @@
+        /// Initializes a new instance of the <see cref="GpioController"/> class that will use the specified numbering scheme and driver.
+        /// </summary>
+        /// <param name="numberingScheme">The numbering scheme used to represent pins provided by the controller.</param>
+        public AnalogController(PinNumberingScheme numberingScheme)
```
consider making this protected

krwq replied:
although here I think it doesn't make any difference - match visiblility with other similar classes in .NET

pgrawehr replied:
Class is abstract, so a public ctor is useless - changed.

krwq commented on src/devices/Arduino/samples/DebugLogStream.cs:
```diff
@@ -0,0 +1,90 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
```
can we simplify the sample and print directly to console rather than having this class?

pgrawehr replied:
Removed this class for now. It is only useful for very specific low-level testing.

krwq commented on src/devices/Arduino/samples/CharacterDisplay.cs:
```diff
@@ -0,0 +1,48 @@
+
+namespace Iot.Device.Arduino.Sample
+{
+    internal sealed class CharacterDisplay : IDisposable
```
consider creating a separate demo for each of those (i.e. one simple project per file)

krwq commented on src/devices/Arduino/samples/Arduino.sample.cs:
```diff
@@ -0,0 +1,506 @@
+
+        private static bool Menu(ArduinoBoard board)
+        {
+            Console.WriteLine("Hello I2C and GPIO on Arduino!");
+            Console.WriteLine("Select the test you want to run:");
+            Console.WriteLine(" 1 Run I2C tests with a BMP280");
+            Console.WriteLine(" 2 Run GPIO tests with a simple led blinking on GPIO6 port");
+            Console.WriteLine(" 3 Run polling button test on GPIO2");
+            Console.WriteLine(" 4 Run event wait test event on GPIO2 on Falling and Rising");
+            Console.WriteLine(" 5 Run callback event test on GPIO2");
+            Console.WriteLine(" 6 Run PWM test with a LED dimming on GPIO6 port");
+            Console.WriteLine(" 7 Blink the LED according to the input on A1");
+            Console.WriteLine(" 8 Read analog channel as fast as possible");
+            Console.WriteLine(" 9 Run SPI tests with an MCP3008 (experimental)");
+            Console.WriteLine(" 0 Detect all devices on the I2C bus");
+            Console.WriteLine(" H Read DHT11 Humidity sensor on GPIO 3 (experimental)");
+            Console.WriteLine(" X Exit");
```
I'd suggest simplifying this and creating separate projects instead (at least for sensors, GPIO is ok to be joint into one)

pgrawehr replied:
I think that won't make things simpler, especially also for maintaining them.

krwq commented on src/devices/Arduino/samples/Arduino.sample.cs:
```diff
@@ -0,0 +1,506 @@
+    /// </summary>
+    internal class Program
+    {
+        /// <summary>
+        /// Main entry point
+        /// </summary>
+        /// <param name="args">Unused</param>
+        public static void Main(string[] args)
+        {
+            string portName = "COM4";
+            if (args.Length > 1)
+            {
+                portName = args[1];
+            }
```
```suggestion
        public static void Main(string[] args)
        {
            if (args.Length == 0)
            {
                Console.WriteLine("Usage: Arduino.sample <PortName>");
                Console.WriteLine("i.e.: Arduino.sample COM4");
                return;
            }

            string portName = args[0];
```

krwq replied:
the comment above Main is not really useful

krwq commented on src/devices/Arduino/samples/Arduino.Monitor.csproj:
```diff
@@ -0,0 +1,30 @@
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net5.0</TargetFramework>
+    <RootNamespace>Iot.Device.Arduino.Sample</RootNamespace>
+	<EnableDefaultItems>false</EnableDefaultItems>
```
indent off

krwq commented on src/devices/Arduino/samples/Arduino.Monitor.cs:
```diff
@@ -0,0 +1,300 @@
+                buttonClicked.WaitOne(sleeptime);
+            }
+
+            hardwareMonitor.Dispose();
+            disp.Output.Clear();
+            disp.Dispose();
+            bmp?.Dispose();
+            gpioController.Dispose();
```
1. consider using `using` statement instead
2. Why are you null checking for bmp but not for others?

pgrawehr replied:
The others are non-nullable. The bmp sensor is optional.

joperezr commented on src/devices/Arduino/Arduino.csproj:
```diff
@@ -0,0 +1,19 @@
+  <ItemGroup>
+    <ProjectReference Include="$(MainLibraryPath)System.Device.Gpio.csproj" />
+    <ProjectReference Include="..\Common\CommonHelpers.csproj" />
+    <PackageReference Include="System.IO.Ports" Version="4.7.0" />
```
Can we move these two package references to eng\Versions.props as properties so that we reference the same version across the repo?

pgrawehr replied:
Sure, they actually already are there.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,413 @@
+        /// <param name="portName">Port name. On Windows, this is usually "COM3" or "COM4" for an Arduino attached via USB.
+        /// On Linux, possible values include "/dev/ttyAMA0", "/dev/serial0", "/dev/ttyUSB1", etc.</param>
+        /// <param name="baudRate">Baudrate to use. It is recommended to use at least 115200 Baud.</param>
+        public ArduinoBoard(string portName, int baudRate)
```
I wonder if we should just have a constructor that takes the SerialPort object directly.

pgrawehr replied:
I think this is easier to use, but whatever you like. You can always use `serialPort.BaseStream` as argument to the other ctor, though.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,413 @@
+
+            if (_serialPort != null)
+            {
+                _serialPort.Dispose();
```
I suppose the answer is no, but will this throw if the _dataStream is disposed first and it was fetched using this port?

pgrawehr replied:
No. I would have seen that.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,413 @@
+
+            if (_firmata != null)
+            {
+                _firmata.OnError -= FirmataOnError;
```
NIT: you can just make sure that on the Firmata class you clear out the OnError handler when calling Dispose instead of doing it here.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,413 @@
+                _firmata.Dispose();
+            }
+
+            _firmata = null;
```
NIT: move this up on inside the null check if block.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,413 @@
+        /// <summary>
+        /// Default pin numbering scheme of this board. Always <see cref="PinNumberingScheme.Logical"/>, because the Arduino has only one pin numbering scheme.
+        /// </summary>
+        public PinNumberingScheme DefaultPinNumberingScheme
```
NIT: make this internal or private? I don't really see the use of this when it always return the same and no one really seems to call it.

pgrawehr replied:
As noted elsewhere already: This will later be defined by the base class, where it is not necessarily constant.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,413 @@
+        /// <summary>
+        /// Attach to this event to retrieve log messages. The Exception argument may be null if it is only an informational message.
+        /// </summary>
+        public event Action<string, Exception?>? LogMessages;
```
NIT: Consider changing this to use EventHandler instead of declaring your delagates:
```c#
public event EventHandler<LoggingEventArgs>? LogMessages;
```
Based on the framework design guidelines: [https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/event](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/event)

pgrawehr replied:
I'll completely rewire this once the logging stuff goes in (possibly removing it altogether from the API of this class).

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,413 @@
+        /// <param name="comPorts">List of com ports. Can be used with <see cref="SerialPort.GetPortNames"/>.</param>
+        /// <param name="baudRates">List of baud rates to test. <see cref="CommonBaudRates"/>.</param>
+        /// <returns>A board, already open and initialized. Null if none was found.</returns>
+        public static ArduinoBoard? FindBoard(IEnumerable<string> comPorts, IEnumerable<int> baudRates)
```
Whats the use case of having this overload public? Won't people always want to call the parameterless one? when would somebody would want to manually pass in ports and baudRates?

pgrawehr replied:
There are multiple possible reasons to use this: You might want to scan only a set of ports, because you know which ports are certainly wrong and should not be probed (see also comment). And you might be using a non-default baudrate, or more likely, know the correct baudrate. I use this overload in tests to find the board, but with a known baudrate.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,413 @@
+        /// </summary>
+        public static List<int> CommonBaudRates()
+        {
+            return new List<int>()
```
Just curious, where did we get these from?

pgrawehr replied:
Knowledge. The ones below 115200 are well known values, above that it's more or less arbitrary.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,413 @@
+        /// Initialize the board connection. This must be called before any other methods of this class.
+        /// </summary>
+        /// <exception cref="NotSupportedException">The Firmata firmware on the connected board is too old.</exception>
+        public virtual void Initialize()
```
I still see this guy public, I thought we said we were going to add a _initialized field and check on every method that needs it and call Initialize if it hasn't been called yet. That way you a) can remove Initialize from the public surface area and b) you avoid a very common pitfall folks will have when they forget to call initialize.

pgrawehr replied:
I forgot about this... But I'd rather have to keep it public and throw if _initialzed is false, to force correct usage. This should be explicly called, to make sure exceptions are handled in one single place. I think it would be bad if we get exceptions from initialization errors from all possible methods.

joperezr replied:
All exceptions will really come from Initialize in the callstack though. This pattern of lazy initialization is used elsewhere in the repo, for example: https://github.com/dotnet/iot/blob/b5bb684bbf00ebd266018cfffde3dcbde26f0538/src/System.Device.Gpio/System/Device/Spi/Devices/UnixSpiDevice.cs#L44

I think it is not great to design an API that we know will require a method to be called first, but doesn't call it and instead expects the user to have called it. The only argument that could be made is that you get a minor perf hit since you now  have to check on every action if it has been initialized or not, but seems like you are going to do that anyway in order to throw the exception.

pgrawehr replied:
The main reason for separating the ctor from init is actually testability. If you want to write unit tests for individual functions with Mocks, you do not want to write up the mocks for the initialization sequence in each test case. (There's a few other things missing here to do be able to do that, but that's up for another PR).

joperezr replied:
We shouldn't be designing Api surface for testing purposes, at least not when that means affecting the usability of the types. Primary focus should be usability, for tests we could use other techniques like InternalsVisibleTo or making methods virtual to override them, or just using Moq framework. For this particular case, we could move the actual implementation that doesn't call Initialize to an internal method, then use InternalsVisibleTo for the tests and use Moq to change the implementation and call the internal method directly without calling Initialize.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,413 @@
+        /// <param name="serialPortStream">A stream to an Arduino/Firmata device</param>
+        public ArduinoBoard(Stream serialPortStream)
+        {
+            _dataStream = serialPortStream;
```
Make sure that you check that the passed in stream isn't null. If you don't app will fail during initialize.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,413 @@
+
+            Log($"Firmware version on board is {FirmwareVersion}");
+
+            _firmata.QueryCapabilities();
```
Looks like this isn't actually doing anything with the state other than testing if communication is working right? If this is just for debugging purposes, it might be better to add this code and the one that actually logs all of the capabilities  behind #Debug flag to not affect perf. You don't have to do this, I'm just mentioning it for you to consider.

pgrawehr replied:
No, this is filling the capability list (see next line). So this is required at startup. The log part will later be routed trough the log library, with a LogLevel of debug or info.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,413 @@
+        /// <summary>
+        /// Returns the list of capabilities per pin
+        /// </summary>
+        public ReadOnlyCollection<SupportedPinConfiguration> SupportedPinConfigurations
```
NIT: it is usually better practice to return interfaces when appropriate, for example here you could do `IReadOnlyList<SupportedPinConfiguration>` and above where you are returning BaudRates you can use `IEnumerable<int>` or `IList<int>`

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,413 @@
+        /// </summary>
+        /// <param name="pinNumberingScheme">Pin numbering scheme to use for this controller</param>
+        /// <returns>An instance of GpioController, using an Arduino-Enabled driver</returns>
+        public virtual GpioController CreateGpioController(PinNumberingScheme pinNumberingScheme)
```
I thought you said that only logical pin numbering was supported by firmata driver right?

pgrawehr replied:
True, but this is supposed to be overriding the generic method from the (future) base class.

joperezr replied:
I see, in that case I would remove it from this PR since you are going to change the signature anyway when you add the board class as you will need to change virtual => override. I would try to break any ties that this PR has with the Board one for now since the Board one can still change and we don't want to add dependencies here that we may change later.

pgrawehr replied:
I can do that now, but as I already said: I do expect that there will be changes for these adjustments anyway.

joperezr replied:
I'll leave it up to you, but I think that the way to go about it is to start with the minimal, and then add the rest once the board PR comes in.

pgrawehr replied:
Removed the overload with an argument for now.

joperezr commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,413 @@
+        /// <exception cref="NotSupportedException">The Bus number is not 0, or the SPI component has not been enabled in the firmware.</exception>
+        public virtual SpiDevice CreateSpiDevice(SpiConnectionSettings settings)
+        {
+            if (settings.BusId != 0)
```
NIT: check settings is not null, similar above for i2c

joperezr commented on src/devices/Arduino/ArduinoAnalogController.cs:
```diff
@@ -0,0 +1,101 @@
+            // This method is called with the logical pin numbering (input pin A0 is 0, A1 is 1, etc)
+            // but the SetPinMode method operates on the global numbers
+            int fullNumber = ConvertLogicalNumberingSchemeToPinNumber(pinNumber);
+            _board.Firmata.SetPinMode(fullNumber, SupportedMode.ANALOG_INPUT);
```
>_board [](start = 12, length = 6)

Keep in mind that it is possible that _board is already disposed, in which case a bunch of things in here might not work appropriately. That is the same case for the rest of the devices (ArduinoGpioControllerDriver, ArduinoI2cDevice, ArduinoSpiDevice). I think that ArduinoBoard should have a internal bool disposed, and all of these devices that were created using it should throw whenever that has been disposed already.

pgrawehr replied:
Just tested again: That works already. The `_board.Firmata` property throws an ObjectDisposedException when the board is disposed.

joperezr replied:
I thought you were setting Firmata to null, in any case I think it would be better to not rely on Firmata implementation as that may change, and instead make sure that _board also throws ObjectDisposedException when trying to access it or one of its fields.

pgrawehr replied:
`Firmata` is a property with a backing field. Only the backing field is set to null.

joperezr commented on src/devices/Arduino/ArduinoAnalogController.cs:
```diff
@@ -0,0 +1,101 @@
+    internal class ArduinoAnalogController : AnalogController
+    {
+        private readonly ArduinoBoard _board;
+        private readonly ReadOnlyCollection<SupportedPinConfiguration> _supportedPinConfigurations;
```
Same feedback here, try to use Interfaces instead of concrete classes.

krwq commented on src/devices/Arduino/ArduinoAnalogController.cs:
```diff
@@ -0,0 +1,101 @@
+                }
+            }
+
+            throw new InvalidOperationException($"Pin A{logicalPinNumber} is not existing");
```
```suggestion
            throw new InvalidOperationException($"Pin A{logicalPinNumber} does not exist");
```

krwq commented on src/devices/Arduino/ArduinoAnalogController.cs:
```diff
@@ -0,0 +1,101 @@
+    {
+        private readonly ArduinoBoard _board;
+        private readonly IReadOnlyList<SupportedPinConfiguration> _supportedPinConfigurations;
+        private readonly Dictionary<int, ValueChangedEventHandler> _callbacks;
```
you don't seem to be using the callbacks, consider removing

pgrawehr commented:
TODO
- Fix Initialisation
- Introduce _disposed flag
- Check thread safety on ArduinoBoard
- Consider moving lock in FirmataDevice to a single place

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,514 @@
+        /// <param name="board">[Out] Returns the board reference. It is already initialized.</param>
+        /// <returns>True on success, false if no board was found</returns>
+        public static bool TryFindBoard(IEnumerable<string> comPorts, IEnumerable<int> baudRates,
+#if !NETCOREAPP2_1
```
@joperezr  can we do similar trick with attributes we do in dotnet/runtime so we can avoid `#if`s like this?

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,514 @@
+        /// </summary>
+        /// <exception cref="NotSupportedException">The Firmata firmware on the connected board is too old.</exception>
+        /// <exception cref="TimeoutException">There was no answer from the board</exception>
+        public virtual void Initialize()
```
I do not like having APIs like this but I'm ok with it for now. Consider making it protected though

pgrawehr replied:
Good catch. I was actually discussing just this with @joperezr, but in the end forgot to change the actual function.

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,514 @@
+        {
+            lock (_initializationLock)
+            {
+                if (_initialized)
```
I'd recommend copying this if just before the lock as well and remove all `if (!_initialized) Initialize();` from the code

pgrawehr replied:
Yea, that was needlessly complicated.

krwq commented on src/devices/Arduino/ArduinoBoard.cs:
```diff
@@ -0,0 +1,514 @@
+        {
+            get
+            {
+                if (!_initialized)
+                {
+                    Initialize();
+                }
```
see comment above, consider moving this logic into Initialize (currently it's there inside of the lock, but consider doing this check before and inside of the lock)

krwq commented on src/devices/Arduino/ArduinoGpioControllerDriver.cs:
```diff
@@ -0,0 +1,305 @@
+                    ret = PinMode.Input;
+                    break;
+                default:
+                    ret = PinMode.Input; // TODO: Return "Unknown"
```
consider adding `Debug.Fail`

krwq replied:
(or logging)

krwq commented on src/devices/Arduino/ArduinoGpioControllerDriver.cs:
```diff
@@ -0,0 +1,305 @@
+            }
+
+            var pinConfig = _supportedPinConfigurations.FirstOrDefault(x => x.Pin == pinNumber);
+            if (pinConfig == null || !pinConfig.PinModes.Contains(firmataMode))
+            {
+                return false;
+            }
+
+            return true;
```
```suggestion
            return pinConfig != null && pinConfig.PinModes.Contains(firmataMode);
```

krwq commented on src/devices/Arduino/ArduinoGpioControllerDriver.cs:
```diff
@@ -0,0 +1,305 @@
+
+        protected override PinValue Read(int pinNumber)
+        {
+            return _arduinoBoard.Firmata.GetDigitalPinState(pinNumber);
+        }
+
+        protected override void Write(int pinNumber, PinValue value)
+        {
+            _arduinoBoard.Firmata.WriteDigitalPin(pinNumber, value);
```
nit: I find the naming on the `Firmata` really weird, consider: `ReadDigitalPin`/`WriteDigitalPin`.

krwq commented on src/devices/Arduino/ArduinoGpioControllerDriver.cs:
```diff
@@ -0,0 +1,305 @@
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                lock (_callbackContainersLock)
+                {
+                    _callbackContainers.Clear();
+                }
+
+                _arduinoBoard.Firmata.DigitalPortValueUpdated -= FirmataOnDigitalPortValueUpdated;
+            }
+
+            base.Dispose(disposing);
```
consider adding null checks and nulling out all fields so GC can collect

krwq replied:
note: for nullable annotations just suppress messages with `!` since we do not care about correctness after dispose

pgrawehr replied:
The fields are read-only, so I suppose a Clear() should mostly do it. Several of the values (i.e. the `_arduinoBoard` reference) aren't owned by this class, so nulling it has no effect anyway.

krwq commented on src/devices/Arduino/FirmataCommandSequence.cs:
```diff
@@ -0,0 +1,60 @@
+﻿using System;
```
copyright header

krwq commented on src/devices/Arduino/README.md:
```diff
@@ -0,0 +1,106 @@
+# SPI, GPIO and I2C drivers for Arduino
```
you might want to add Firmata somewhere in the name since this will be on the device listing

krwq commented on src/devices/Arduino/README.md:
```diff
@@ -0,0 +1,106 @@
+
+## Device family
+
+This binding remotely controls different Arduino boards directly from PC Software. It provides support for accessing GPIO ports as well as I2C devices, SPI devices, PWM output and analog input. The Arduino is remote controlled by individual commands from the PC, the entire program will run on the PC, and not on the Arduino, so the connection cannot be removed while the device is being used. 
```
```suggestion
This binding remotely controls Arduino boards directly from PC Software. It provides support for accessing GPIO ports as well as I2C devices, SPI devices, PWM output and analog input. The Arduino is remote controlled by individual commands from the PC, the entire program will run on the PC, and not on the Arduino, so the connection cannot be removed while the device is being used. 
```

krwq commented on src/devices/Arduino/README.md:
```diff
@@ -0,0 +1,106 @@
+
+## Desktop Requirements
+
+In order to have an Arduino board working with the PC, you need to Install the Arduino IDE together with the drivers for your board type. If you get a simple sketch uploaded and running (such as the blinking LED example) you are fine to start. If you're new to the Arduino world, read the introduction at https://www.arduino.cc/en/Guide for a quick start. The explanations below assume you have the  ...
```
technically this is not needed. Once the Arduino is flashed you don't need Arduino IDE

pgrawehr replied:
True, but you might still need to install the USB driver. I have never tried installing that without the IDE.

pgrawehr replied:
Added a sentence saying that this is (usually) a one-time task.

krwq replied:
looks good now

krwq commented on src/devices/Arduino/README.md:
```diff
@@ -0,0 +1,106 @@
+                    return;
+                }
+
+                ArduinoBoard board = new ArduinoBoard(port.BaseStream);
```
consider showcasing simpler constructor (portname, baudRate) so sample is shorter

krwq commented on src/devices/Arduino/README.md:
```diff
@@ -0,0 +1,106 @@
+
+All communication is routed trough the USB cable immitating a serial port with a limited bandwith. Therefore, some not insignificant delays are to be expected when sending commands or retrieving data. Communicating with sensors which have time-critical behavior will most likely not work reliably for this reason and the standard bindings provided for these won't work. This includes sensors like th ...
+
+For the moment this binding supports GPIO, Analog In, SPI, I2C and DHT on all platforms. Here is the list of tested features:
+
+- [x] SPI master support for Windows 64/32
+- [x] I2C master support for Windows 64/32
+- [x] Basic GPIO support for Windows 64/32
+- [x] Advanced GPIO support for Windows 64/32
+- [x] Analog input support on Windows 64/32
+- [ ] SPI support for MacOS 
+- [ ] I2C support for MacOS
+- [ ] GPIO support for MacOS
+- [ ] SPI support for Linux 64
+- [ ] I2C support for Linux 64
+- [ ] GPIO support for Linux 64
```
I'd recommend removing this. You haven't used any interop so this should technically just work. If not for some mysterious reason then someone will create an issue

krwq commented on src/devices/Arduino/tests/ApiBehaviorTests.cs:
```diff
@@ -0,0 +1,53 @@
+﻿using System;
```
copyright header

krwq commented on src/devices/Common/System/Device/Analog/AnalogInputPin.cs:
```diff
@@ -0,0 +1,171 @@
+        /// The reference voltage level to convert raw values into voltages.
+        /// Some boards (i.e. the ADS111x series) always return an absolute voltage. Then this value is meaningless.
+        /// </summary>
+        public double VoltageReference
```
why not use Voltage from Units.Net?

pgrawehr replied:
Because I just forgot this one.

krwq commented:
@pgrawehr looks good to me with some comments

krwq commented on src/devices/Arduino/ArduinoGpioControllerDriver.cs:
```diff
@@ -0,0 +1,301 @@
+                    break;
+                default:
+                    _arduinoBoard.Log($"Unexpected pin mode found: {mode}. Is the pin not set to GPIO?");
+                    ret = PinMode.Input; // TODO: Return "Unknown"
```
nit: TODO is no longer needed since you added logging

krwq commented on src/devices/Arduino/README.md:
````diff
@@ -0,0 +1,81 @@
+
+Basic start:
+```
+            // Portname is "COM3", "COM4" on Windows, "/dev/ttyUSB0" or similar on linux
+            Console.WriteLine($"Connecting to Arduino on {portName}");
+			ArduinoBoard board = new ArduinoBoard("COM3", 115200);
+			try
+			{
+				board.LogMessages += BoardOnLogMessages; // Get log messages
+				Console.WriteLine($"Connection successful. Firmware version: {board.FirmwareVersion}, Builder: {board.FirmwareName}");
+				// Add code that uses the board here.
+			}
+			catch (TimeoutException x)
+			{
+				Console.WriteLine($"No answer from board: {x.Message} ");
+			}
+			finally
+			{
+				port.Close();
+				board?.Dispose();
+			}
````
- consider moving the indentation so it starts with 0
- indent of first two lines is off
- set syntax to csharp

krwq commented:
@pgrawehr I think https://github.com/dotnet/iot/pull/1039#pullrequestreview-595131842 got missed other than that LGTM

krwq reviewed (approved):
I'd recommend to fix the missed comment but nothing blocking to merge this from my perspective. Thanks for this great contribution!

krwq commented:
:shipit:
